Add instructions to allow GPDB development using Vagrant

pateljm opened the pull request:
As we started playing with GPDB at the University of Wisconsin, Navneet Potti (@navsan), Nabarun Nag (@nabarunnag) and I felt that it was too hard for new developers to get GPDB compiled and running. We sat down and tried out a few approaches and came up with a minimal Vagrant-based recipe, which is in this pull request. Hopefully other developers can benefit from these instructions too.

ivannovick commented:
Hi Jignesh,

Looks like you need to change your membership in the organization greenplum-db to "public" as per: https://help.github.com/articles/publicizing-or-hiding-organization-membership/

ivannovick commented on DEV_SETUP.md:
```diff
@@ -0,0 +1,326 @@
+# Developer's guide for GPDB
```
Is this in the root directory?

pateljm replied:
It is right now, so developers can see it upfront. Any other suggestions for a location? It could go under the vagrant directory, but then that would be hard to find for new developers. 

Also @ivannovick I made my Pivotal membership public on github. Thanks for pointing that out!

danielgustafsson replied:
I would prefer if this document was placed as `vagrant/README.md` and instead add a note in the main `README.md` about Vagrant setup and where to find the documentation.

pateljm replied:
Good idea @danielgustafsson! I just moved it to the /vagrant directory and added a note in the main README.md file as you suggested. I think this also better addresses @ivannovick original comment.

danielgustafsson commented on DEV_SETUP.md:
```diff
@@ -0,0 +1,326 @@
+
+First, list the Postgres processes by typing in (a guest terminal) the following 
+command: `ps ax | grep postgres`. You should see a list that looks something 
+like: ![Postgres processes](/vagrant/pictures/gpdb_processes.png)
```
Is is possible to render the `ps` output in text format directly in the document in some way for those reading the document in a terminal? If at all possible in a good way, retaining sensible line widths is of course important for readability.

pateljm replied:
Dear Daniel: I played around with tying to embed in colored text, but wasn't very successful as the lines are long. The image is high-res and clickable, so if it doesn't render nicely, the reader can click on it. I added a quick note in the instructions to alert the reader to this. Thanks for the feedback!

danielgustafsson replied:
Fair enough, thanks for taking a stab at it.

danielgustafsson commented:
Thanks for working on this!

ivannovick commented:
@schubert can you help get the bot to pick up the CLA now?

d commented on vagrant/Vagrantfile:
```diff
@@ -0,0 +1,42 @@
+  # Base box: Centos-7 box
+  # NOTE: Over time the VMI below may become outdated, so may need to be 
+  #       substituted with a more recent VMI
+  config.vm.box = "boxcutter/centos72"
```
This is awesome, I love `Vagrant`. Although, isn't it more idiomatic to make a vagrant box that has all the things we are installing in the provisioner? In fact, how do you like it if you document such dependencies in the `packer` script for said vagrant box.

pateljm replied:
Dear Jesse: Great point. I was trying to get the developer started with minimal steps and wanted to reduce the number of tools that they need. I was also trying to find a good balance between full automation and showing the developer some details (that they will likely need to do serious development). I did think of `packer` and for the reasons above went with the instructions here.

davecramer commented:
I had a look at this. For some reason initdb fails if you attempt to do make cluster from /gpdb/gpAux/gpdemo with "FATAL","XX000","could not open file ""pg_xlog/000000010000000000000001""

 Even moving gpdemo to a non shared directory fails as well with "FATAL","21000","Greenplum Database number of segment databases cannot be 0",

pateljm commented:
@davecramer Thanks for the comment re. making gpdemo work. See the changes that just got checked in. Let me know if you see additional issues.

danielgustafsson commented:
How do we best proceed with this and #267, can/should they be combined? I'm not very experienced with Vagrant but I would really like to see this get in to help lower the barriers to adoption. Ping @davecramer & @pateljm.

pateljm commented:
How about we consider closing this PR first (as I think was first), and then consider what the purpose of #267 is (I'm not quite sure). @davecramer?

davecramer commented:
I think the solution is to create multiple vagrant files. Not everyone wants to work with redhat. I prefer debian, which is what my vagrant does. I need a new PR, as I created a new branch for it. I'll rework mine so it does not conflict. But I'd suggest naming the Vagrant file Vagrant.fedora and Vagrant.debian in my case

davecramer commented on vagrant/Vagrantfile:
```diff
@@ -0,0 +1,42 @@
+  end
+
+  # Make the gpdb folder will be visible as /gpdb in the virtual machine 
+  config.vm.synced_folder "../../gpdb", "/gpdb"
```
I don't think it's a good thing to assume that the cloned directory is gpdb. It would be better to just use ../.

pateljm replied:
Good idea. Changing now as per your suggestion.

davecramer commented on vagrant/README.md:
````diff
@@ -0,0 +1,405 @@
+vm.overcommit_memory = 2
+```
+Hit the escape key, then type `:wq` to finish saving the file. Verify that you
+see the changes that you made by typing in `cat /etc/sysctl.conf`.
````
I'd suggest putting this in /etc/sysctl.d/gpdb.conf

pateljm replied:
Excellent suggestion! Making that change now.

davecramer commented on vagrant/README.md:
```diff
@@ -0,0 +1,405 @@
+see the changes that you made by typing in `cat /etc/sysctl.conf`.
+
+##6: Reboot the Guest OS
+We have to reboot the guest OS to allow the changes to the `sysctl.conf`
```
It's not necessary to reboot the OS. sudo sysctl -p /etc/sysctl.conf or whatever file you put the changes in will make the changes take effect immediately

davecramer commented:
I redid my PR#322 so that it would not conflict with this one. I'd suggest moving this one into vagrant/centos to allow others to live happily beside it.

pateljm commented:
@davecramer Thank you for the excellent suggestions! I have created a vagrant/centos directory, moved the Vagrantfile there, and updated the instructions to reflect the new directory. 

I'd suggest that after this PR is closed, you can create another PR to create debian/Vagrantfile and update the README.md file to simply state in Step #3 that the reader could substituted centos with debian and follow the remaining instructions.

davecramer commented:
closed with https://github.com/greenplum-db/gpdb/commit/4cccd743e301971463178a1454c7e30ca649c245

danielgustafsson commented:
These commits should've been squashed into a single commit before pushing to master.

pateljm commented:
Thanks @davecramer and @danielgustafsson for all your hard work on this PR. It was a pleasure working with you!
